File: java/com/tinyutils/StringUtils.java
package com.tinyutils;

import java.util.regex.Pattern;

/** Immutable string transformation utilities. Every method is pure. */
public final class StringUtils {

    private static final Pattern NON_WORD = Pattern.compile("[^\\w\\s-]");
    private static final Pattern SPACES = Pattern.compile("[-\\s]+");

    private StringUtils() {}

    /** URL-safe slug from free-form text. */
    public static String slugify(String text) {
        if (text == null || text.isEmpty()) return "";
        String s = text.toLowerCase().trim();
        s = NON_WORD.matcher(s).replaceAll("");
        s = SPACES.matcher(s).replaceAll("-");
        return s.replaceAll("^-+|-+$", "");
    }

    /** Truncate at word boundary. */
    public static String truncate(String text, int maxLen, String suffix) {
        if (text == null || text.length() <= maxLen) return text;
        int space = text.substring(0, maxLen).lastIndexOf(' ');
        return (space > 0 ? text.substring(0, space) : text.substring(0, maxLen)) + suffix;
    }

    public static String truncate(String text, int maxLen) { return truncate(text, maxLen, "..."); }
}

File: typescript/string-utils.ts
/**
 * String transformation utilities (TypeScript).
 *
 * Pure functions — no side effects, no mutation.
 */

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const NON_WORD = /[^\w\s-]/g;
const SPACES = /[-\s]+/g;
const CAMEL_FIRST = /([A-Z]+)([A-Z][a-z])/g;
const CAMEL_SECOND = /([a-z0-9])([A-Z])/g;

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/** Convert free-form text into a URL-safe slug. */
export function slugify(text: string): string {
  const cleaned = text.toLowerCase().trim().replace(NON_WORD, "");
  return cleaned.replace(SPACES, "-").replace(/^-+|-+$/g, "");
}

/** Truncate text at the nearest word boundary without exceeding maxLength. */
export function truncate(
  text: string,
  maxLength: number = 80,
  suffix: string = "..."
): string {
  if (text.length <= maxLength) return text;

  const candidate = text.slice(0, maxLength);
  const boundary = candidate.lastIndexOf(" ");
  return boundary > 0 ? candidate.slice(0, boundary) + suffix : candidate + suffix;
}

/** Convert camelCase or PascalCase to snake_case. */
export function camelToSnake(name: string): string {
  return name
    .replace(CAMEL_FIRST, "$1_$2")
    .replace(CAMEL_SECOND, "$1_$2")
    .toLowerCase();
}

/** Reverse the order of whitespace-delimited words. */
export function reverseWords(text: string): string {
  return text.split(/\s+/).reverse().join(" ");
}

/** Convert dash-/underscore-delimited text to PascalCase. */
export function toPascal(text: string): string {
  return text
    .split(/[-_]/)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join("");
}
